components/uniset/src/builder.rs

```diff
@@ -0,0 +1,801 @@
+use std::{char, cmp, cmp::Ordering, ops::RangeBounds};
+
+use crate::{uniset::UnicodeSet, utils::deconstruct_range};
+
+/// UnicodeSet builder wrapper
+///
+/// Provides exposure to builder functions and conversion to UnicodeSet
+#[derive(Default)]
+pub struct UnicodeSetBuilder {
+    intervals: Vec<u32>,
+}
+
+impl UnicodeSetBuilder {
+    /// Returns empty UnicodeSetBuilder
+    pub fn new() -> UnicodeSetBuilder {
+        UnicodeSetBuilder { intervals: vec![] }
+    }
+
+    /// Returns a UnicodeSet and consumes the UnicodeSetBuilder
+    pub fn build(self) -> UnicodeSet {
+        UnicodeSet::from_inversion_list(self.intervals).unwrap()
+    }
+
+    /// Adds the range to the end of the builder
+    fn append(&mut self, start: u32, end: u32) {
+        self.intervals.push(start);
+        self.intervals.push(end);
+    }
+
+    /// Adds the range to the beginning of the builder in a single memory move
+    fn prepend(&mut self, start: u32, end: u32) {
+        let mut new = vec![start, end];
+        new.append(&mut self.intervals);
+        self.intervals = new;
+    }
+
+    /// Add the range to the UnicodeSetBuilder
+    ///
+    /// Accomplishes this through binary search for the start and end indices and merges intervals
+    /// in between with inplace memory. Performs `O(1)` operation if adding to end of list, and `O(N)` otherwise,
+    /// where `N` is the number of endpoints.
+    fn add(&mut self, start: u32, end: u32) {
+        if start >= end {
+            return;
+        }
+        if self.intervals.is_empty() {
+            self.append(start, end);
+            return;
+        }
+        let last_ind = self.intervals.len() - 1;
+        if self.intervals[last_ind - 1] <= start {
+            if self.intervals[last_ind] < start {
+                self.append(start, end);
+            } else {
+                self.intervals[last_ind] = cmp::max(end, self.intervals[last_ind]);
+            }
+        } else if end <= self.intervals[1] {
+            if end < self.intervals[0] {
+                self.prepend(start, end);
+            } else {
+                self.intervals[0] = cmp::min(self.intervals[0], start);
+            }
+        } else {
+            let s = self.intervals.binary_search(&start);
+            let e = self.intervals.binary_search(&end);
+            let mut s_index = s.unwrap_or_else(|x| x);
+            let mut e_index = e.unwrap_or_else(|x| x);
+            let s_even: bool = s_index % 2 == 0;
+            let e_even: bool = e_index % 2 == 0;
+            let s_equals_e = s_index == e_index;
+
+            if s_equals_e && s_even && e.is_err() {
+                self.intervals.splice(s_index..e_index, vec![start, end]);
```

Comment:
Avoid the `vec!`: I think `&[start, end]` should work, but if not, make a temporary variable (2-element array) and pass it in as a slice.

Reply:
agree and changed, but same as above regarding syntax